### DISCUSSION TITLE
How to connect time travel client desktop app to android/ios emulator app?

### DISCUSSION THREAD

**Discussion by plusmobileapps** [badoo/MVIKotlin#289] [Q&A] [answered]
I was looking over the [documentation](https://arkivanov.github.io/MVIKotlin/time_travel.html) related to the time travel client and cannot seem to figure out how to connect the client to the android and iOS emulator. 

## Android 

I could connect it to a physical android device using the ip address from the device's about settings and default port of `6379` for the time travel server.  However cannot seem to figure out how to connect it to either the sample or the app with the ip address `localhost` or `127.0.0.1` with a connection refused error? any thoughts? 

## iOS

For the ios emulator I could connect the time travel client to the sample todo app with `127.0.0.1`, although since I'm new to ios I was wondering how to export the time travel server to a library so it can be started from any app like [the sample](https://github.com/arkivanov/MVIKotlin/blob/master/sample/todo-app-ios/todo-app-ios/AppDelegate.swift#L15)?

**Comment by arkivanov** [comment, marked as answer]
For Android it is better to connect via ADB. The idea plugin is able to connect only via ADB. The desktop client app has settings with both ADB and IP/port modes. When ADB mode is used, the client executes an `adb` command to forward TCP port (6379) between the app and the host. Then it connects as usual to localhost:6379. If not using the ADB mode, you will have to forward the port manually via ADB.

For iOS, you need to add `mvikotlin-timetravel` module as API dependency and export it to the framework as usual: https://github.com/arkivanov/MVIKotlin/blob/4e0624946fe24a2fc47ecbfeb35a6fecaf09f709/sample/todo-darwin-umbrella/build.gradle.kts#L27

**Reply by plusmobileapps** [reply]
Ahh I see, surprised I could connect to a physical device without ADB. Good to know that the plugin only works through adb and was finally able to connect it to the android emulator! Does the plugin have to use port 6379 or is there an option like in the client to switch the port? 

Also got it to connect to the iOS emulator after making the time travel dependency an `api()` and exporting it. 

Overall I really love the tool, I think it's really powerful in its capabilities and thank you for the help🚀  

Was wondering if you think we should add some more documentation around the feature as there was a little bit of digging to get it up and running? I could probably help out with that need be

**Reply by arkivanov** [reply]
> Does the plugin have to use port 6379 or is there an option like in the client to switch the port?

The plugin does not have the settings dialog yet. I added it to the client app because it was easy with Compose. But the plugin uses Swing currently, so I did not have enough time to work on this. I would appreciate a contribution here as well. Filed #291 just in case.

> Was wondering if you think we should add some more documentation around the feature

Agree, there is a room for improvement here indeed. Filed #292.

**Reply by arkivanov** [reply]
> surprised I could connect to a physical device without ADB

It is actually expected! You can even connect remotely, as long as server's IP address and port are reachable. E.g. it should work fine in a home/office local network. ADB is only used to forward ports, so the emulator's port becomes accessible.

**Reply by plusmobileapps** [reply]
I could take care of #292, #291 might need a little more effort as I'm very inexperienced with swing as well but could be a fun project.

**Reply by arkivanov** [reply]
Thanks @plusmobileapps , contributions are very welcome.
